### PR TITLE
Refactro goalcanvas

### PR DESCRIPTION
**Pull request by muhtasimhafiz**
#### Description

Please include a short description of the suggested change and the reasoning behind the approach you have chosen.

#### Link to ticket

Please add a link to the GitHub issue being addressed by this change.

### Type

[] Fix
[] Feature
[x] Cleanup 

#### Screenshot of the result

If your change affects the user interface, you should include a screenshot of the result with the pull request.

### CODE REVIEW

File: app/Domain/Canvas/Repositories/Canvas.php [hunk 1]
```diff
@@ -282,7 +282,7 @@ public function getAllCanvas($projectId, $type = null): false|array
          * @param $canvasId
          * @return array|false
          */
-        public function getSingleCanvas($canvasId): false|array
+        public function getSingleCanvas($canvasId)
```
Comment (marcelfolaron) [inline app/Domain/Canvas/Repositories/Canvas.php:285]: Why did you remove that? This is the return value definition and always good to have.

File: app/Domain/Goalcanvas/Controllers/BigRock.php [hunk 2]
```diff
@@ -68,7 +68,7 @@ public function get($params): Response
                 $bigrock = array("id"=>'', "title" => "", "prpojectId" => "", "author" => '');
             }
 
-
+            dd($bigrock);
```
Comment (marcelfolaron) [inline app/Domain/Goalcanvas/Controllers/BigRock.php:71]: Please remove

File: app/Domain/Goalcanvas/Controllers/ShowCanvas.php [hunk 3]
```diff
@@ -265,10 +64,41 @@ public function run()
             $this->tpl->assign('allCanvas', $allCanvas);
             $this->tpl->assign('canvasItems', $this->goalService->getCanvasItemsById($currentCanvasId));
             $this->tpl->assign('users', $this->projectService->getUsersAssignedToProject(session("currentProject")));
+        
+            return $this->tpl->display(static::CANVAS_NAME . 'canvas.showCanvas');
```
Comment (marcelfolaron) [inline app/Domain/Goalcanvas/Controllers/ShowCanvas.php:68]: Since this is the goal canvas you can change "static::CANVAS_NAME" to 'goalcanvas.showCanvas'

File: app/Domain/Goalcanvas/Controllers/ShowCanvas.php [hunk 4]
```diff
@@ -265,10 +64,41 @@ public function run()
...
+                    $result = $this->goalService->mergeCanvas($_POST['canvasid'] ?? -1, $_POST['mergeCanvasId'] ?? -1);
+                    break;
+                case 'importCanvas':
+                    $result = $this->goalService->importCanvas($_FILES['canvasfile'] ?? null);
+                    break;
+            }
+        
+            if ($result['success']) {
+                $this->tpl->setNotification($result['message'], 'success');
+                return Frontcontroller::redirect(BASE_URL . '/' . static::CANVAS_NAME . 'canvas/showCanvas/');
```
Comment (marcelfolaron) [inline app/Domain/Goalcanvas/Controllers/ShowCanvas.php:96]: See above. Canvas name can be renamed to just be canvas

File: app/Domain/Goalcanvas/Repositories/Goalcanvas.php [hunk 5]
```diff
@@ -151,13 +153,15 @@ public function getSingleCanvas($canvasId): false|array
                 ORDER BY zp_canvas.title, zp_canvas.created";
 
             $stmn = $this->db->database->prepare($sql);
-            $stmn->bindValue(':canvasId', $canvasId, PDO::PARAM_STR);
+            $stmn->bindValue(':canvasName', static::CANVAS_NAME . 'canvas', PDO::PARAM_STR);
```
Comment (marcelfolaron) [inline app/Domain/Goalcanvas/Repositories/Goalcanvas.php:156]: Can replace this with goalcanvas

File: app/Domain/Goalcanvas/Services/Goalcanvas.php [hunk 6]
```diff
@@ -3,13 +3,27 @@
...
      */
     class Goalcanvas
     {
         private GoalcanvaRepository $goalRepository;
+        private ?Projects $projectService;
+
+        protected ?Language $language;
+        protected Mailer $mailer;
+
+        protected const CANVAS_NAME = 'goal';
```
Comment (marcelfolaron) [inline app/Domain/Goalcanvas/Services/Goalcanvas.php:24]: You can remove that and just replace the CANVAS_NAME instances to goal

File: app/Domain/Goalcanvas/Services/Goalcanvas.php [hunk 7]
```diff
@@ -202,5 +223,488 @@ public function pollForUpdatedGoals(): array|false
...
+                'title' => $params['canvastitle'],
+                'author' => session("userdata.id"),
+                'projectId' => session("currentProject"),
+            ];
+            $currentCanvasId = $this->goalRepository->addCanvas($values);
+
+            $this->sendCanvasNotification('notification.board_created', $values['title']);
+
+            session(["current" . strtoupper(static::CANVAS_NAME) . "Canvas" => $currentCanvasId]);
+            return $this->setRedirect($result, BASE_URL . '/' . static::CANVAS_NAME . 'canvas/showCanvas/');
```
Comment (marcelfolaron) [inline app/Domain/Goalcanvas/Services/Goalcanvas.php:395]: The service layer should ideally not perform any redirects. If the service is called via API the redirect would break the api response. 

I think in this case the method could just return either an error message or the id of the new canvas.

File: app/Domain/Goalcanvas/Services/Goalcanvas.php [hunk 8]
```diff
@@ -202,5 +223,488 @@ public function pollForUpdatedGoals(): array|false
...
+        }
+
+        private function handleNewCanvas($params, $result): array
+        {
+            if (!isset($params['canvastitle']) || empty($params['canvastitle'])) {
+                return $this->setNotification($result, 'notification.please_enter_title', 'error');
+            }
+
+            if ($this->goalRepository->existCanvas(session("currentProject"), $params['canvastitle'])) {
+                return $this->setNotification($result, 'notification.board_exists', 'error');
```
Comment (marcelfolaron) [inline app/Domain/Goalcanvas/Services/Goalcanvas.php:382]: Please move template notifications into the controller. The service should just return success or failure messages. (Maybe with failure errors/exceptions) 
So in this case you could throw a "ElementExistsException()" (just create a new exception type into core/exceptions. The controller can then decide what to do with that exception. (aka show a notification)

File: app/Domain/Goalcanvas/Services/Goalcanvas.php [hunk 9]
```diff
@@ -202,5 +223,488 @@ public function pollForUpdatedGoals(): array|false
...
+                $result = $this->handleImportCanvas($params, $result);
+            }
+
+            return $result;
+        }
+
+        private function handleNewCanvas($params, $result): array
+        {
+            if (!isset($params['canvastitle']) || empty($params['canvastitle'])) {
+                return $this->setNotification($result, 'notification.please_enter_title', 'error');
```
Comment (marcelfolaron) [inline app/Domain/Goalcanvas/Services/Goalcanvas.php:378]: This should throw a MissingParameterException()

File: app/Domain/Ideas/Templates/ideaDialog.blade.php [hunk 10]
```diff
@@ -0,0 +1,163 @@
+@extends($layout)
+
+@section('content')
+    @php
+        $id = '';
```
Comment (marcelfolaron) [inline app/Domain/Ideas/Templates/ideaDialog.blade.php:5]: This should be moved into the controller

**Review by marcelfolaron — CHANGES_REQUESTED** [review]
Overall really good update.
I am not fully done with the review but there are 2 big items we should address:

1. The service layer should not create template notifications. Since the service is also our API those notifications would not be available to API endpoints. Instead the services should throw exceptions if something goes wrong/is missing. Or it should return true or return the id of a newly created item etc. 

2. Similar to the notifications the service layers should not do any redirects. Same reason with regards to the API. API endpoints would get redirected to a UI endpoint and fail.

File: app/Domain/Goalcanvas/Services/Goalcanvas.php (hunk 9 above)
Comment (marcelfolaron) [reply app/Domain/Goalcanvas/Services/Goalcanvas.php:378]: `throw new MissingParameterException()`

File: app/Domain/Goalcanvas/Controllers/Dashboard.php [hunk 11]
```diff
@@ -48,283 +49,41 @@ public function init(
...
-            $this->tpl->assign('users', $this->projectService->getUsersAssignedToProject(session("currentProject")));
-
-            if (!isset($_GET['raw'])) {
-                return $this->tpl->display(static::CANVAS_NAME . 'canvas.dashboard');
-            }
-        }
+         public function get($params): Response
+         {
+
+             $result = $this->goalService->handleDashboardGetRequest($params);
```
Comment (marcelfolaron) [inline app/Domain/Goalcanvas/Controllers/Dashboard.php:56]:
```
try {
$result = $this->goalService->handleDashboardGetRequest($params);
}catch(MissingParameterException $e) {
  $this->tpl->setNotification
}catch() {
}finally {
$result = [];
}
```

File: app/Domain/Goalcanvas/Controllers/BigRock.php (hunk 2 above)
Comment (marcelfolaron) [reply app/Domain/Goalcanvas/Controllers/BigRock.php:71]: Why did you resolve the comment but not remove the line?

Comment (muhtasimhafiz) [reply app/Domain/Goalcanvas/Controllers/BigRock.php]: i resolved it before pushing the code, I resolved it locally

File: app/Domain/Goalcanvas/Controllers/BigRock.php [hunk 12]
```diff
@@ -66,7 +66,6 @@ public function get($params): Response
                 $bigrock = array("id"=>'', "title" => "", "prpojectId" => "", "author" => '');
             }
 
-
```
Comment (marcelfolaron) [inline app/Domain/Goalcanvas/Controllers/BigRock.php:69]: Few items missing from this file:
- No repo calls (they should all be service calls. If there is no service, just create a wrapper service method
- CANVAS_NAME is still being used

File: app/Domain/Goalcanvas/Controllers/EditCanvasComment.php [hunk 13]
```diff
@@ -101,6 +101,7 @@ public function get($params)
             }
 
             $this->tpl->assign('comments', $comments);
+            $this->tpl->assign('id', $canvasItem['id']);
```
Comment (marcelfolaron) [inline app/Domain/Goalcanvas/Controllers/EditCanvasComment.php:104]: Few items missing from this file:
- No repo calls (they should all be service calls. If there is no service, just create a wrapper service method
- CANVAS_NAME is still being used

File: app/Domain/Goalcanvas/Controllers/EditCanvasItem.php [hunk 14]
```diff
@@ -201,8 +201,8 @@ public function post($params): Response
                             'parent' => $params['parent'] ?? null,
                             "id" => $params['itemId'],
                             'kpi' => $params['kpi'] ?? '',
-                            'startDate' => format(value: $params['startDate'], fromFormat: FromFormat::UserDateStartOfDay)->isoDateTime(),
-                            'endDate' => format(value: $params['endDate'], fromFormat: FromFormat::UserDateEndOfDay)->isoDateTime(),
+                            'startDate' => !empty($params['startDate'])?format(value: $params['startDate'], fromFormat: FromFormat::UserDateStartOfDay)->isoDateTime():'',
```
Comment (marcelfolaron) [inline app/Domain/Goalcanvas/Controllers/EditCanvasItem.php:204]: Few items missing from this file:
- No repo calls (they should all be service calls. If there is no service, just create a wrapper service method
- CANVAS_NAME is still being used

**Review by marcelfolaron — CHANGES_REQUESTED** [review]
Please review all the controllers and make sure that 

- No repo calls (they should all be service calls. If there is no service, just create a wrapper service method
- CANVAS_NAME is still being used and should be replaced. 
- Remove commented out code snippets
- Add doc comments to service
